pkg/virt-api/rest/instancetype_test.go

```diff
@@ -0,0 +1,255 @@
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+
+	kubevirtcore "kubevirt.io/api/core"
+	v1 "kubevirt.io/api/core/v1"
+	instancetypev1alpha1 "kubevirt.io/api/instancetype/v1alpha1"
+	"kubevirt.io/client-go/generated/kubevirt/clientset/versioned/fake"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/testutils"
+)
+
+var _ = Describe("Instancetype expansion subresources", func() {
+	const (
+		vmName      = "test-vm"
+		vmNamespace = "test-namespace"
+	)
+
+	var (
+		vmClient            *kubecli.MockVirtualMachineInterface
+		virtClient          *kubecli.MockKubevirtClient
+		instancetypeMethods *testutils.MockInstancetypeMethods
+		app                 *SubresourceAPIApp
+
+		request  *restful.Request
+		recorder *httptest.ResponseRecorder
+		response *restful.Response
+
+		vm *v1.VirtualMachine
+	)
+
+	BeforeEach(func() {
+		ctrl := gomock.NewController(GinkgoT())
+		vmClient = kubecli.NewMockVirtualMachineInterface(ctrl)
+		virtClient = kubecli.NewMockKubevirtClient(ctrl)
+		virtClient.EXPECT().GeneratedKubeVirtClient().Return(fake.NewSimpleClientset()).AnyTimes()
+		virtClient.EXPECT().VirtualMachine(vmNamespace).Return(vmClient).AnyTimes()
+
+		instancetypeMethods = testutils.NewMockInstancetypeMethods()
+
+		app = NewSubresourceAPIApp(virtClient, 0, nil, nil)
+		app.instancetypeMethods = instancetypeMethods
+
+		request = restful.NewRequest(&http.Request{})
+		recorder = httptest.NewRecorder()
+		response = restful.NewResponse(recorder)
+		response.SetRequestAccepts(restful.MIME_JSON)
+
+		vm = &v1.VirtualMachine{
+			TypeMeta: metav1.TypeMeta{},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      vmName,
+				Namespace: vmNamespace,
+			},
+			Spec: v1.VirtualMachineSpec{
+				Template: &v1.VirtualMachineInstanceTemplateSpec{
+					Spec: v1.VirtualMachineInstanceSpec{
+						Domain: v1.DomainSpec{},
+					},
+				},
+			},
+		}
+	})
+
+	testCommonFunctionality := func(callExpandInstancetypeApi func(vm *v1.VirtualMachine) *httptest.ResponseRecorder, expectedStatusError int) {
+		It("should return unchanged VM, if no instancetype and preference is assigned", func() {
+			vm.Spec.Instancetype = nil
+
+			recorder := callExpandInstancetypeApi(vm)
+			Expect(recorder.Code).To(Equal(http.StatusOK))
+
+			responseVm := &v1.VirtualMachine{}
+			Expect(json.NewDecoder(recorder.Body).Decode(responseVm)).To(Succeed())
+			Expect(responseVm).To(Equal(vm))
+		})
+
+		It("should fail if VM points to nonexistent instancetype", func() {
+			instancetypeMethods.FindInstancetypeSpecFunc = func(_ *v1.VirtualMachine) (*instancetypev1alpha1.VirtualMachineInstancetypeSpec, error) {
+				return nil, fmt.Errorf("instancetype does not exist")
+			}
+
+			vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+				Name: "nonexistent-instancetype",
+			}
+
+			recorder := callExpandInstancetypeApi(vm)
+			_ = ExpectStatusErrorWithCode(recorder, expectedStatusError)
+		})
+
+		It("should fail if VM points to nonexistent preference", func() {
```

Comment:
As above.

Reply:
As above again, we don't protect against the removal of preferences between VM creation and start so this is actually valid.